doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+            if (Value *New_val = replaceZero(Inst, C)) {
+              Inst.setOperand(i, New_val);
+              modified = true;
+            } else {
+              //dbgs() << "ObfuscateZero: could not rand pick a variable for replacement\n";
+            }
+          }
+        }
+      }
+      registerInteger(Inst);
+    }
+
+
+  bool isValidCandidateInstruction(Instruction &Inst) {
+    if (isa<GetElementPtrInst>(&Inst)) {
+      // dbgs() << "Ignoring GEP\n";
+      return false;
+    } else if (isa<SwitchInst>(&Inst)) {
+      // dbgs() << "Ignoring Switch\n";
+      return false;
+    } else if (isa<CallInst>(&Inst)) {
+      // dbgs() << "Ignoring Calls\n";
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+Pretty easy, no? Well the hard part is that this kind of problems is almost impossible to anticipate unless you know all the LLVM instructions.
+The only solution to find this id to run your passes on big projects, see where it crashes and find out why.
+
+Your code should now be the same as the ObfuscateZero pass.
+And since I don't want to dump all the code on this page (again) from now on we are going to use the ObfuscateZero pass for our tests.
+
+Tests, tests and more tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+I hope this last part made you understand that validation is crutial before using your obfuscations in prod.
+For ObfuscateZero we used the ``llvm-lit`` testing tool, shipped with LLVM.
+This tool runs the tests you specify with a particular syntax. Take a look in the files in the ``$PASSDIR/tests/ObfuscateZero`` folder to learn how to use it.
+
+For ObufuscateZero we have two types of tests:
+    * Simple tests checking if the pass actually does what we want and doesn't crash in some tricky cases (GEP :p)
+    * The validation scripts (*.sh files). Those files download the sources from openssl and zlib, compile them with our pass and run their validation suite. If the project compiles without error *and* passes the validation suite, we can suppose that our pass doesn't introduce bugs.
+
+If you want to run the tests you first need to install `pip <https://pypi.python.org/pypi/pip>`_ (python-pip). Then use it to install lit:
+
+.. code:: bash
+
+    >$ pip install lit
+
+Then go to ``$PASSDIR/build`` and run :
+.. code:: bash
+
+    >$ cmake -DLLVM_ROOT=path/to/your/llvm/build ..
+    >$ make check
+
+This will run the ``ObfuscateZero`` tests, which you can modify to test your pass. But it's going to take some time.
+To validate ``ObfuscateZero`` we also compiled a C++ code since some constructs are not present when compiling from C.
+However the test file has not been shipped in the git.
+
+Bye Bye
+=======
+
+This tutorial was just an introduction to writing LLVM passes and using them for obfuscation.
+There are many more funny things to do to make your code very annoying for reverse engineers.
+I hope this will help you get started.
```

Comment:
But rember, if you choose the quick and easy path as Vader did - you will become an agent of evil.